tests/system/Debug/ExceptionsTest.php

```diff
@@ -32,9 +33,55 @@ final class ExceptionsTest extends CIUnitTestCase
 
     protected function setUp(): void
     {
+        parent::setUp();
+
         $this->exception = new Exceptions(new ExceptionsConfig(), Services::request(), Services::response());
     }
 
+    /**
+     * @requires PHP >= 8.1
```

Comment:
What about one of the 8.0 deprecations? This one would probably be easy to do with an anonymous function:

> Deprecate required parameters after optional parameters in function/method signatures

Reply:
It seems this is not workable. PHP analyzes the function signature at compile time so our error handler is not registered yet.
```
Deprecated: Optional parameter $str declared before required parameter $num is implicitly treated as a required parameter in C:\Users\P\Desktop\Web Dev\CodeIgniter4\tests\system\Debug\ExceptionsTest.php on line 55

[Call Stack]
....

PHPUnit 9.5.25 #StandWithUkraine

Runtime:       PHP 8.1.11
Configuration: C:\Users\P\Desktop\Web Dev\CodeIgniter4\phpunit.xml.dist

F......                                                             7 / 7 (100%)

Time: 00:00.497, Memory: 48.00 MB

There was 1 failure:

1) CodeIgniter\Debug\ExceptionsTest::testDeprecationsOnPhp80DoNotThrow
Failed asserting that logs have a record of message containing "Optional parameter $str declared before required parameter $num" with level "error".
Failed asserting that false is true.

C:\Users\P\Desktop\Web Dev\CodeIgniter4\system\Test\CIUnitTestCase.php:374
C:\Users\P\Desktop\Web Dev\CodeIgniter4\tests\system\Debug\ExceptionsTest.php:56
```

Reply:
It's happening at zend_compile.c
https://github.com/nikic/php-src/blob/9b4601a65c84d0e35ef70fc64473904cb5a5d77c/Zend/zend_compile.c#L5848-L5856

Reply:
https://github.com/php/php-src/issues/8529

Reply:
Nice sleuthing! Let's stick with what you have; the rest of the 8.0 deprecations look even more complicated.